src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs

```diff
@@ -1395,83 +1390,128 @@ internal bool StartsWithLoop(int upperBoundLowestValue = 1)
-                        return _right.GetPrefixSequence(pref.Add(_left._set), lengthBound - 1);
+                        // Either a path didn't have a next set as supported by this algorithm, or the next set was not equal
+                        yield break;
                     }
-                    return pref;
-
-                case SymbolicRegexKind.Or:
-                case SymbolicRegexKind.And:
-                    {
-                        Debug.Assert(_alts is not null);
-
-                        SymbolicRegexSet<S>.Enumerator enumerator = _alts.GetEnumerator();
-                        bool movedNext = enumerator.MoveNext();
-                        Debug.Assert(movedNext, "Expected a minimum of one element");
-                        ImmutableList<S> altsPrefix = enumerator.Current.GetPrefixSequence(ImmutableList<S>.Empty, lengthBound);
+                }
+                // At this point all paths generated equal next sets
+                yield return next;
+                if (done)
+                {
+                    // Some path had no continuation, end the prefix
+                    yield break;
+                }
+                else
+                {
+                    Debug.Assert(paths.Count == 0, "Not all paths were considered for next set.");
+                    paths.AddRange(nextPaths);
+                    nextPaths.Clear();
+                }
+            }
+        }
 
-                        while (!altsPrefix.IsEmpty && enumerator.MoveNext())
+        private bool GetNextPrefixSet(ref List<SymbolicRegexNode<S>> paths, ref HashSet<SymbolicRegexNode<S>> nextPaths, ref bool done, out S set)
+        {
+            while (paths.Count > 0)
+            {
+                SymbolicRegexNode<S> node = paths[paths.Count - 1];
+                paths.RemoveAt(paths.Count - 1);
+                switch (node._kind)
+                {
+                    case SymbolicRegexKind.Singleton:
+                        Debug.Assert(node._set is not null);
+                        set = node._set;
+                        done = true; // No continuation, done after the next set
+                        return true;
+                    case SymbolicRegexKind.Concat:
+                        Debug.Assert(node._left is not null && node._right is not null);
+                        if (node._left._kind == SymbolicRegexKind.Singleton)
                         {
-                            ImmutableList<S> p = enumerator.Current.GetPrefixSequence(ImmutableList<S>.Empty, lengthBound);
-                            int prefix_length = altsPrefix.TakeWhile((x, i) => i < p.Count && x.Equals(p[i])).Count();
-                            altsPrefix = altsPrefix.RemoveRange(prefix_length, altsPrefix.Count - prefix_length);
+                            nextPaths.Add(node._right);
+                            Debug.Assert(node._left._set is not null);
+                            set = node._left._set;
```

Comment:
Here I thought one could use more generally node._startSet, to cover some cases when 
the concat does not start with a singleton (but e.g. (a|A) rather than [aA]) but I'm not exactly sure if that matters, it might already be optimized away by the constructor (or, rather, (a|A)  **should** be optimized to [aA] at construction time).
Unfortunately, the mimimum/maximum length information is not part of info that you could just check that 
the length of _left here is 1 and then use the _startSet. This is not important though.

Reply:
Oh right, using _startSet is a great idea. With the current algorithm the prefix does need to end there though, since for patterns like (aa|AA)b the prefix needs to be aab (+ignore case). The current algorithm can only track a path that only descends further into the syntax tree. This could be improved by, instead of having a single SymbolicRegexNode for each path, having a stack of nodes where the top one is followed until it ends and the path ends when the stack empties.